### PR TITLE
Refactor block indexing

### PR DESCRIPTION
**Pull request by osipov-mit**

### CODE REVIEW

File: idea/data-storage/src/gear/gear.service.ts [hunk 1]
```diff
@@ -5,36 +5,42 @@ import { filterEvents } from '@polkadot/api/util';
...
+import { sleep } from '../utils/sleep';
+import config from '../config/configuration';
 
-const { gear } = configuration();
+const gearConfig = config().gear;
 
 const MAX_RECONNECTIONS = 10;
 let reconnectionsCounter = 0;
-
+interface BlockParams {
```
Comment (Almazatun) [inline idea/data-storage/src/gear/gear.service.ts:32]: Please create a `types` folder in `gear` folder to contain the types of this service

File: idea/data-storage/src/gear/gear.service.ts [hunk 2]
```diff
@@ -88,71 +95,88 @@ export class GearService {
...
+      if (this.newBlocks.length === 0) {
+        await sleep(1000);
+        continue;
+      }
+      yield this.newBlocks.shift();
+    }
+  }
+
+  private async indexBlocks() {
+    for await (const { hash, number } of this.blocksGenerator()) {
```
Comment (Almazatun) [inline idea/data-storage/src/gear/gear.service.ts:120]: You have the value `number` in the `loop`. Can you rename the value to a more meaningful name to recognize what is the `magic` number value?

File: idea/data-storage/src/gear/gear.service.ts [hunk 3]
```diff
@@ -88,71 +95,88 @@ export class GearService {
...
+      if (this.lastBlockNumber === undefined) {
+        this.logger.log(`Block processing started with ${number}.`);
+      } else if (number === this.lastBlockNumber) continue;
+      else if (number - 1 !== this.lastBlockNumber) {
+        this.logger.warn(
+          // eslint-disable-next-line max-len
+          `Some blocks have been missed. The last processed block is ${this.lastBlockNumber} but the current is ${number}`,
+        );
+        for (let n = this.lastBlockNumber + 1; n < number; n++) {
```
Comment (Almazatun) [inline idea/data-storage/src/gear/gear.service.ts:129]:
```suggestion
        for (let blockNumber = this.lastBlockNumber + 1; n < number; blockNumber++) {
```

File: idea/data-storage/src/gear/gear.service.ts [hunk 4]
```diff
@@ -88,71 +95,88 @@ export class GearService {
...
+    },
+    [Keys.ProgramChanged]: async (data: any) =>
+      await this.programService.setStatus(data.id, this.genesis, data.programStatus),
+    [Keys.MessagesDispatched]: (data: any) => this.messageService.setDispatchedStatus(data),
+    [Keys.UserMessageRead]: (data: any) => this.messageService.updateReadStatus(data.id, data.reason),
+    [Keys.CodeChanged]: (data: any) => this.codeService.updateCodes([data]),
+  };
+
+  private async handleEvents(block: SignedBlockExtended, timestamp: number, hash: HexString): Promise<void> {
+    for (const {
```
Comment (Almazatun) [inline idea/data-storage/src/gear/gear.service.ts:166]: Can you please refactor this part of the code a little bit?

```javascript
const methods = Object.keys(Keys);
const filteredBlockEvents = block.events.filter(({ event: { method } }) => methods.includes(method))
for (const { event: { data, method } } of filteredBlockEvents) {
      try {
        const eventData = eventDataHandlers[method](data as GenericEventData);

        if (eventData === null) continue;

        eventData.blockHash = hash;
        await this.eventHandlers[method](eventData, timestamp);
      } catch (error) {
        this.logger.warn({ method, data: data.toHuman() });
        console.error(error);
      }
    }
    ````

File: idea/common/src/enums/code-status.enum.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
-export enum CodeStatus  {
+export enum CodeStatus {
   ACTIVE = 'Active',
   INACTIVE = 'Inactive',
+  UNKNOWN = 'Unknown',
```
Comment (Almazatun) [inline idea/common/src/enums/code-status.enum.ts:4]: Please do not forget to check is need it to create a database migration `file` by `code entity`

**Comment by Almazatun** [comment]
Please also do not forget to add a description of what you did if it's possible 👍

**Review by Almazatun — APPROVED** [review]

File: idea/common/src/enums/code-status.enum.ts (hunk 5 above)
Comment (osipov-mit) [reply idea/common/src/enums/code-status.enum.ts:4]: btw, thanks

**Pull request merged**